tempesta_fw/http_parser.c

```diff
@@ -685,29 +686,86 @@ do {									\
-		 msg->flags, msg->content_length);			\
-	/* RFC 2616 4.4: firstly check chunked transfer encoding. */	\
+/* Process according RFC 7230 3.3.3 */
+#define TFW_HTTP_INIT_REQ_BODY_PARSING()				\
+__FSM_STATE(RGen_BodyInit) {						\
+	/*								\
+	 * TODO: If both msg->h_tbl[TFW_HTTP_HDR_TRANSFER_ENCODING]	\
+	 * and msg->h_tbl[TFW_HTTP_HDR_CONTENT_LENGTH] are present,	\
+	 * then issue and error or remove "Content-Length:" header.	\
+	 */								\
+	/*								\
+	 * TODO: msg->h_tbl[TFW_HTTP_HDR_TRANSFER_ENCODING]		\
+	 * is present and "chunked" coding is the last in the list.	\
+	 */								\
 	if (msg->flags & TFW_HTTP_CHUNKED)				\
-		__FSM_MOVE_nofixup(RGen_Body);				\
-	/* Next we check content length. */				\
-	if (msg->content_length						\
-	    && !(msg->flags & TFW_HTTP_VOID_BODY))			\
-	{								\
+		__FSM_MOVE_nofixup(RGen_BodyStart);			\
+	/*								\
+	 * TODO: msg->h_tbl[TFW_HTTP_HDR_TRANSFER_ENCODING]		\
+	 * is present and there's NO "chunked" coding.			\
+	 * Must send 400 (Bad Request) and close the connection.	\
+	 */								\
+	if (msg->content_length) {					\
 		parser->to_read = msg->content_length;			\
-		__FSM_MOVE_nofixup(RGen_Body);				\
+		__FSM_MOVE_nofixup(RGen_BodyStart);			\
 	}								\
-	/* There is no body at all. */					\
+	/* There is no body. */						\
 	msg->body.flags |= TFW_STR_COMPLETE;				\
 	r = TFW_PASS;							\
 	FSM_EXIT();							\
-} while (0)
+}
+
+/* Process according RFC 7230 3.3.3 */
+#define TFW_HTTP_INIT_RESP_BODY_PARSING()				\
+__FSM_STATE(RGen_BodyInit) {						\
+	TfwHttpResp *resp = (TfwHttpResp *)msg;				\
+									\
+	TFW_DBG3("parse msg body: flags=%#x content_length=%lu\n",	\
+		 msg->flags, msg->content_length);			\
+									\
+	/* There's no body. */						\
+	/* TODO: Add (req == CONNECT && resp == 2xx) */			\
+	if (((resp->status > 100) && (resp->status < 200))		\
+	    || (resp->status == 204) || (resp->status == 304)		\
+	    || (msg->flags & TFW_HTTP_VOID_BODY))			\
+	{								\
+		msg->body.flags |= TFW_STR_COMPLETE;			\
+		r = TFW_PASS;						\
+		FSM_EXIT();						\
+	}								\
+	/*								\
+	 * TODO: Both msg->h_tbl[TFW_HTTP_HDR_TRANSFER_ENCODING]	\
+	 * and msg->h_tbl[TFW_HTTP_HDR_CONTENT_LENGTH] are present.	\
+	 * Issue and error or remove "Content-Length:" header.		\
+	 */								\
+	/*								\
+	 * TODO: msg->h_tbl[TFW_HTTP_HDR_TRANSFER_ENCODING]		\
+	 * is present and "chunked" coding is the last in the list.	\
+	 */								\
+	if (msg->flags & TFW_HTTP_CHUNKED)				\
+		__FSM_MOVE_nofixup(RGen_BodyStart);			\
+	/*								\
+	 * TODO: msg->h_tbl[TFW_HTTP_HDR_TRANSFER_ENCODING]		\
+	 * is present and there's NO "chunked" coding.			\
+	 * Process the body until the connection is closed.		\
+	 */								\
+	if (msg->content_length) {					\
+		parser->to_read = msg->content_length;			\
+		__FSM_MOVE_nofixup(RGen_BodyStart);			\
+	}								\
+	/* Process the body until the connection is closed. */		\
+	tfw_http_msg_set_data(msg, &msg->body, p);			\
```

Comment:
As `p` points to the last EOL character, not the first body one, we can't do `tfw_http_msg_set_data` here. Need to `__FSM_MOVE_nofixup` first.